File: pdoo/java/Deepspace/SpaceCity.java
/**
 * @author Manuel Gachs Ballegeer
 */
package Deepspace;

import java.util.ArrayList;

/**
 * @brief Clase ciudad espacial
 */
public class SpaceCity extends SpaceStation {
    private SpaceStation base;
    private ArrayList<SpaceStation> collaborators = new ArrayList<>();
    
    SpaceCity(SpaceStation _base,ArrayList<SpaceStation> rest) {
        super(_base);
        base = _base;
        collaborators = rest;
    }
    
    public ArrayList<SpaceStation> getCollaborators() { return collaborators; }
    
    @Override
    public float fire() {
        float fire_power = super.fire();
        for (SpaceStation s: collaborators)
            fire_power += s.fire();
        return fire_power;
    }
    
    @Override
    public float protection() {
        float shield_power = super.protection();
        for (SpaceStation s: collaborators)
            shield_power += s.protection();
        return shield_power;
    }
    
    @Override
    public Transformation setLoot(Loot loot) {
        super.setLoot(loot);
        return Transformation.NOTRANSFORM;
    }
    
    @Override
    public String toString() {
        String aux = "-----CIUDAD ESPACIAL-----\nCiudad base:\n";
        aux += super.toString() + "\nColaboradores:\n";
        for (SpaceStation s: collaborators)
            aux += s.toString();
        return aux;
    }
    
    @Override
    SpaceCityToUI getUIversion() { return new SpaceCityToUI(this); }
}

File: pdoo/ruby/spacecity.rb
#encoding:utf-8

require_relative "spacestation"

module Deepspace

  class SpaceCity<SpaceStation

    attr_reader :collaborators

    def initialize(base,rest)
      newCopy(base)
      @base = base
      @collaborators = rest
    end

    def fire
      fire_power = super
      collaborators.each do |station|
        fire_power += station.fire
      end
      return fire_power
    end

    def protection
      shield_power = super
      collaborators.each do |station|
        shield_power += station.protection
      end
      return shield_power
    end

    def setLoot(loot)
      super
      return Transformation::NOTRANSFORM
    end

    def to_s
      aux = "-----CIUDAD ESPACIAL-----\nBASE\n"
      aux += super
      aux += "\nCOLABORADORES\n"
      collaborators.each do |station|
        aux += station.to_s
      end
      return aux
    end
  end
end

